core/src/ast/utils.rs

```diff
@@ -21,3 +24,233 @@ pub fn get_doc_lines(attrs: &[Attribute]) -> String {
 
     lines
 }
+
+#[derive(Clone, PartialEq, Eq, Hash, Serialize, Deserialize, Debug)]
+pub struct RustLink {
+    path: Path,
+    typ: DocType,
+}
+
+#[derive(Clone, PartialEq, Eq, Hash, Serialize, Deserialize, Debug)]
+enum DocType {
+    Struct,
+    StructField,
+    Enum,
+    EnumVariant,
+    EnumVariantField,
+    Trait,
+    FnInStruct,
+    FnInEnum,
+    FnInTrait,
+    DefaultFnInTrait,
+    Fn,
+    Mod,
+    Constant,
+    Macro,
+}
+
+pub fn get_rust_link(attrs: &[Attribute]) -> Option<RustLink> {
+    for attr in attrs {
+        if attr.path.to_token_stream().to_string() == "diplomat :: rust_link" {
+            if let Ok(syn::Meta::List(syn::MetaList { nested, .. })) = attr.parse_meta() {
+                if nested.len() == 2 {
+                    if let (
+                        syn::NestedMeta::Meta(syn::Meta::Path(path)),
+                        syn::NestedMeta::Meta(syn::Meta::Path(typ)),
+                    ) = (&nested[0], &nested[1])
+                    {
+                        if let Some(typ) = typ.get_ident() {
+                            return Some(RustLink {
+                                path: Path::from_syn(path),
+                                typ: match typ.to_token_stream().to_string().as_str() {
+                                    "Struct" => DocType::Struct,
+                                    "StructField" => DocType::StructField,
+                                    "Enum" => DocType::Enum,
+                                    "EnumVariant" => DocType::EnumVariant,
+                                    "EnumVariantField" => DocType::EnumVariantField,
+                                    "Trait" => DocType::Trait,
+                                    "FnInStruct" => DocType::FnInStruct,
+                                    "FnInEnum" => DocType::FnInEnum,
+                                    "FnInTrait" => DocType::FnInTrait,
+                                    "DefaultFnInTrait" => DocType::DefaultFnInTrait,
+                                    "Fn" => DocType::Fn,
+                                    "Mod" => DocType::Mod,
+                                    "Constant" => DocType::Constant,
+                                    "Macro" => DocType::Macro,
+                                    x => panic!("Invalid doc type {:?}", x),
+                                },
+                            });
+                        }
+                    }
+                }
+            }
+            panic!("Malformed attribute: {}", attr.to_token_stream());
+        }
+    }
+
+    None
+}
+
+#[derive(Default)]
+pub struct DocsUrlGenerator {
+    base_urls: HashMap<String, String>,
+}
+
+impl DocsUrlGenerator {
+    pub fn with_base_urls(base_urls: HashMap<String, String>) -> Self {
+        Self { base_urls }
+    }
+
+    pub(crate) fn gen_for_rust_link(&self, rust_link: &RustLink) -> String {
```

Comment:
Observation: I think it is odd to make a public class (DocsUrlGenerator) whose primary functionality is hidden in a non-public method.

Reply:
I don't think it's that odd. The main method only needs to construct this and pass it to code that can actually do someting with it. We shouldn't put things in the public API if we don't have to.

Reply:
Yeah, this makes sense

Reply:
I see that `RustLink` is public and `DocsUrlGenerator` is public, but the method that ties them together is private. But, I don't have as strong of a grasp on the API surface here as you do. (My non-blocking review stands)

Reply:
`RustLink` is crate-private as well, even though it has a `pub` modifier, because it's not exported.